packages/swing-store/src/swingStore.js

```diff
@@ -311,69 +383,186 @@ function makeSwingStore(dirPath, forceReset, options) {
+        sql.run();
+        savepoints.length = savepointOrdinal;
+        return;
+      }
     }
-    txnFinish(abort ? lmdbAbort : undefined);
-    return Promise.resolve(txnDone)
-      .then(() => {
-        trace(`${abort ? 'abort' : 'commit'}-tx`);
-      })
-      .finally(() => {
-        txnDone = null;
-        txnFinish = null;
-      });
+    assert.fail(`no such savepoint as "${savepoint}"`);
+  }
+
+  function emitCrankHashes() {
+    // Calculate the resulting crankhash and reset for the next round.
+    const crankhash = crankhasher.finish();
+    resetCrankhash();
+
+    // Get the old activityhash
+    let oldActivityhash = get('activityhash');
+    if (oldActivityhash === undefined) {
+      oldActivityhash = '';
+    }
+
+    // Digest the old activityhash and new crankhash into the new activityhash.
+    const hasher = createSHA256();
+    hasher.add('activityhash');
+    hasher.add('\n');
+    hasher.add(oldActivityhash);
+    hasher.add('\n');
+    hasher.add(crankhash);
+    hasher.add('\n');
+
+    // Store the new activityhash
+    const activityhash = hasher.finish();
+    set('activityhash', activityhash);
+
+    return { crankhash, activityhash };
   }
 
+  function getActivityhash() {
+    return get('activityhash') || '';
+  }
+
+  function endCrank() {
+    assert(inCrank);
+    if (savepoints.length > 0) {
+      sqlReleaseSavepoints.run();
+      savepoints.length = 0;
+    }
+    inCrank = false;
+  }
+
+  const sqlCommit = db.prepare('COMMIT');
+  const sqlCheckpoint = db.prepare('PRAGMA wal_checkpoint(FULL)');
+
   /**
    * Commit unsaved changes.
    */
   async function commit() {
     assert(db);
-    // Commit the stream-store first, before kvStore. We keep the
-    // streams' startPos/endPos in the kvStore, so if we crash after
-    // commitStreamStore() but before the kvstore doCommit(), we'll
-    // wake up with the old startPos/endPos and it's (mostly) as if
-    // the stream entry was never added.
-    commitStreamStore();
-    await doCommit(false);
+    if (db.inTransaction) {
+      sqlCommit.run();
+      sqlCheckpoint.run();
+    }
 
     // NOTE: The kvstore (which used to contain vatA -> snapshot1, and
     //   is being replaced with vatA -> snapshot2)
     //   MUST be committed BEFORE we delete snapshot1.
     //   Otherwise, on restart, we'll consult the kvstore and see snapshot1,
     //   but we'll fail to load it because it's been deleted already.
-    await snapStore.commitDeletes();
+    await (snapStore ? snapStore.commitDeletes() : false);
   }
 
   /**
-   * Close the database, abandoning any changes made since the last commit (if you want to save them, call
-   * commit() first).
+   * Close the database, abandoning any changes made since the last commit (if
+   * you want to save them, call commit() first).
    */
   async function close() {
     assert(db);
-    closeStreamStore();
-    await doCommit(true);
-    await db.close();
+    commit();
+    db.close();
```

Comment:
This changes the behavior on close from `abort` to `commit`.

Reply:
Oh, huh, yeah the previous `true` in `doCommit(true)` triggered an LMDB abort. I must have missed this during review (also we don't have any tests of abort-on-`close()`, and we never use it that way, applications just crash instead of calling `close`). Yeah, we should make this abort.